#13 pre-filter documentation (#36)
nickmoreton/wagtail-wordpress-import: docs/prefilters.md

# Pre-Filtering HTML content

- [Pre-Filtering HTML content](#pre-filtering-html-content)
- [Why use pre-filters](#why-use-pre-filters)
- [Pre-filters](#pre-filters)
  - [Available filters](#available-filters)
  - [Pre-filter running order](#pre-filter-running-order)
  - [Running order:](#running-order)
  - [Configuration](#configuration)
  - [Defaults](#defaults)
  - [Create pre-filters](#create-pre-filters)
- [Pre-filter tools](#pre-filter-tools)
    - [Turn off debugging](#turn-off-debugging)

# Why use pre-filters

> Pre-filters allow transformations to be made on a page's body content before it is converted into blocks. This allows any behavior that affects how the blocks will be generated to be implemented.

Reviewer: **Suggestion**: clarify sentence.

```suggestion
Pre-filters allow transformations to be made on a page's body content before it is converted into blocks. This allows you to implement behaviour that affects how the blocks will be generated.
```

For example, WordPress interprets lines of text separated by two line breaks as separate paragraphs so there's a pre-filter that converts them into separate `"<p>"` tags so that the block generator will convert them into separate paragraphs (using a RichText block).

> The content could also include shortcodes which wordpress uses to represent more structured HTML that might include nested HTML tags like in a FORM. There's a filter that will convert it to a block type/string that can be used to build the block type in the block builder.

Reviewer: **Suggestion**: consistency

```suggestion
The content could also include shortcodes which WordPress uses to represent more structured HTML that might include nested HTML tags like in a FORM. There's a filter that will convert it to a block type/string that can be used to build the block type in the block builder.
```

Reviewer: **Question**: 

> There's a filter that will convert it to a block type/string that can be used to build the block type in the block builder.

Is there? I don't understand this sentence.

Author: I created an issue #52 to remove this paragraph for now

Author: I've been thinking ahead here, you are right there isn't any shortcode handling in place yet.

---

# Pre-filters

Each time the importer processes a page, the body content is transformed by a series of filters.

***Pre-Filters included***

1. linebreaks_wp_filter.py
2. normalize_styles_filter.py
3. fix_styles_filter.py
4. bleach_filter.py

> **The above filters run, by default, in the order above before the content is passed into the page model or parsed to stream field blocks.**

Reviewer: **Suggestion**: consistency

```suggestion
**The above filters run, by default, in the order above before the content is passed into the page model or parsed to StreamField blocks.**
```

## Available filters

-- **filter_linebreaks_wp()**

`wagtail_wordpress_import/prefilters/linebreaks_wp_filter.py`

> Wordpress uses line breaks as a mechanism for vertical spacing of rendered HTML (paragraphs).

Reviewer: ```suggestion
WordPress uses line breaks as a mechanism for vertical spacing of rendered HTML (paragraphs).
```

This filter implements the same process as `wpautop` to convert the raw HTML content into HTML content  with tags around every piece of content. It avoids changing anything that is acceptable. In the end it mostly deals with adding `<p>` tags around text that has no surrounding tag.

-- **filter_normalize_style_attrs()**

`wagtail_wordpress_import/prefilters/normalize_styles_filter.py`

This filter converts inline style attributes to a format we can work with in the next filter because the format of style attributes can be different in some for each tag. It deals with word-case, spacing and closing `;`

-- **filter_fix_styles()**

`wagtail_wordpress_import/prefilters/fix_styles_filter.py`

> This filter applies any inline style to the html tag that is useful in the context of the Wagtail rich text editor. It performs wrapping a tag around a piece of text as well ass applying classes to a tag to represent alignment.

Reviewer: ```suggestion
This filter applies any inline style to the html tag that is useful in the context of the Wagtail rich text editor. It performs wrapping a tag around a piece of text as well as applying classes to a tag to represent alignment.
```

Examples:

> - `font-weigh: bold;` is converted to a `<b>` tag

Reviewer: ```suggestion
- `font-weight: bold;` is converted to a `<b>` tag
```

- `float: left;` is converted to a class of `left-align`

-- **filter_bleach_clean()**

`wagtail_wordpress_import/prefilters/bleach_filter.py`

This filter is run last and removes tags based on a range of `ALLOWED_TAGS`, `ALLOWED_ATTRIBUTES` and `ALLOWED_STYLES` settings in `wagtail_wordpress_import/prefilters/constants.py`.

---

## Pre-filter running order

The order the pre-filters are run will have an impact on the final HTML output.

Each pre-filter takes the output from the previous pre-filter, with the exception of the first pre-filter which receives the raw content. 

> The output of the last pre-filter is used to build the stream field blocks in a later process.

Reviewer: ```suggestion
The output of the last pre-filter is used to build the StreamField blocks in a later process.
```

## Running order: 
It's possible to change the pre-filter running order by altering the order of the `DEFAULT_PREFILTERS` provided by adding `WAGTAIL_WORDPRESS_IMPORT_PREFILTERS` to your  own settings file.

Some pre-filters can be given `OPTIONS` to alter the way the pre-filter behaves.

## Configuration

> You can use the package without providing any configuration for pre-filters. We provide a small number of useful pre-filters that process html content before it's used to build out the stream field blocks.

Reviewer: ```suggestion
You can use the package without providing any configuration for pre-filters. We provide a small number of useful pre-filters that process html content before it's used to build out the StreamField blocks.
```

## Defaults

**Pre-filter Default**:

By default, WAGTAIL_WORDPRESS_IMPORT_PREFILTERS is set to the following:

```python
DEFAULT_PREFILTERS = [
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.linebreaks_wp_filter.filter_linebreaks_wp",
    },
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.normalize_styles_filter.filter_normalize_style_attrs",
    },
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.fix_styles_filter.filter_fix_styles",
    },
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.bleach_filter.filter_bleach_clean",
    },

> ]

Reviewer: **Question**: Are we changing the path names here in this MR, or is that done elsewhere, e.g. in #39?

Author: Yes elsewhere, specifically In the new codebase ticket

```

**To change the configuration:**

> Create you own settings config called `WAGTAIL_WORDPRESS_IMPORT_PREFILTERS` with the FUNCTIONS you need to run, in the order they should be run.

Reviewer: ```suggestion
Create your own settings config called `WAGTAIL_WORDPRESS_IMPORT_PREFILTERS` with the FUNCTIONS you need to run, in the order they should be run.
```

Example: Using custom options in the bleach filter
```python

WAGTAIL_WORDPRESS_IMPORT_PREFILTERS = [
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.linebreaks_wp_filter.filter_linebreaks_wp",
    },
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.normalize_styles_filter.filter_normalize_style_attrs",
    },
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.fix_styles_filter.filter_fix_styles",
    },
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.bleach_filter.filter_bleach_clean",
        "OPTIONS": {
            "ADDITIONAL_ALLOWED_TAGS": ["my-custom-tag"]
        }
    },
]
```

> `my-custom-tag` would be appended to the `ALLOWED_TAGS` in the bleach_filters

Reviewer: **Question**: What are "the bleach_filters"? Is this a typo? I see no explicit mention of it in the repo.

Author: I can clear this up later and added an issue for it #50

or provide your own pre-filter add:

```python
{

>     "FUNCTION": "[app].[file].[method]",

Reviewer: **Suggestion**: Just call it "path to function". We don't want to dictate where users put this method, nor need to tell them how to write Python.

```suggestion
    "FUNCTION": "[path to function]",
```

},
```

> at the position to match the running order you need. `[app].[file].[method]` is the  dotted path to your function.

Reviewer: **Suggestion**: ditto.

```suggestion
at the position to match the running order you need. `[path to function]` is the  dotted path to your function.
```

**If you don't need any of the provided pre-filters you can exclude them from the config**

## Create pre-filters

You can find the provided pre-filters [here](wagtail_wordpress_import/prefilters)

To create your own pre-filter you need to create a function in your app with the following signature:

```python
def filter_func(input_content, options=None):
    """
    params: 
    
    input_content:
    a string of HTML

    options:
    you probably won't need to pass in any options 
    in your own function but the signature requires the 
    parameter to be included. As an example you can pass options
    to `filter_bleach_clean` to allow other HTML tags

    return: 
    most likely a string but depending on your implementation 
    could be any type that you need to work with
    """

    # implement your own transformations/parsing here
    output_content = ...

    return output_content
```

Then you need to include the pre-filter in your own custom config at the position you need it to run. [See here](#defaults)

---

# Pre-filter tools

> If you use the provided `WPImportedPageMixin` it has fields to receive the output from each pre-filter step which can be see on the Debug tab when editing a page. [Read about the mixin](models.md)

Reviewer: ```suggestion
If you use the provided `WPImportedPageMixin` it has fields to receive the output from each pre-filter step which can be seen on the Debug tab when editing a page. [Read about the mixin](models.md)
```

# Pre-Filtering HTML content

- [Pre-Filtering HTML content](#pre-filtering-html-content)
- [Why use pre-filters](#why-use-pre-filters)
- [Pre-filters](#pre-filters)
  - [Available filters](#available-filters)
  - [Pre-filter running order](#pre-filter-running-order)
  - [Running order:](#running-order)
  - [Configuration](#configuration)
  - [Defaults](#defaults)
  - [Create pre-filters](#create-pre-filters)
- [Pre-filter tools](#pre-filter-tools)
    - [Turn off debugging](#turn-off-debugging)

> # Why use pre-filters

Reviewer 2: I think first sentence of this section should answer the question and subsequent sentences provide more detail/examples?

For example:

```
Pre-filters allow transformations to be made on a page's body content before it is converted into blocks. This allows any behaviour that affects how the blocks will be generated to be implemented.

For example, WordPress interprets lines of text separated by two line breaks as separate paragraphs so there's a pre-filter that converts them into separate ``<p>`` tags so that the block generator will convert them into separate paragraph blocks.
```

What do you think?

Author: Cool i changed this

Each item (e.g. page or post) represented in the XML file has content intended for the body of a page that could include WYSIWYG content as well as short-codes.

That content isn't suitable in it's raw form for parsing and creating a range of stream fields or to be added to a single rich text field.

- HTML content can contain inline styles which could be created by WYSIWYG tools or manually entered
- Tables could be present
- Forms could be present
- List could be present
- and more

We do know that the html is going to be suitable for parsing using python built-ins
and other third party packages and be manipulated to better suit the requirements of the import and page creation process in wagtail.

---

# Pre-filters

> Each time the import processes the body content of a page a series of filters are run on the content to transform it.

Reviewer 2: How about:

Each time the importer processes a page, the body content is transformed by a series of filters

Author: i changed this

***Pre-Filters included***

> 1. linebreaks_wp_filter.py: this is a [python implementation](https://gist.github.com/albertsun/1160201/0a1f7d7a509fbcfa580725ed3783a29af51b62b4) of the wordpress function wpauto

Reviewer 2: If we're referencing the py files, should we link to the code? Otherwise, maybe we should use the module paths instead?

Author: I removed the links which I think removes too much information at that point. We have references to the files below anyway.

> 2. normalize_styles_filter.py: using [BeautifulSoup](https://www.crummy.com/software/)

Reviewer 2: Not sure if it's relevant to write how it was implemented here

Author: OK ta. I removed the extra text

3. fix_styles_filter.py: using [BeautifulSoup](https://www.crummy.com/software/BeautifulSoup/)
4. bleach_filter.py: using [bleach](https://github.com/mozilla/bleach)

**The above filters run, by default, in the order above before the content is passed into the page model or parsed to stream field blocks.**

## Available filters

-- **filter_linebreaks_wp()**

`wagtail_wordpress_import/prefilters/linebreaks_wp_filter.py`

Wordpress uses line breaks as a mechanism for vertical spacing of rendered HTML (paragraphs) as an example. 

> This filter implements the same process as `wpauto` to convert the raw html content into html content  with tags around every piece of content. It avoids changing anything that is acceptable. In the end it mostly deals adds `<p>` tags around text that has no surrounding tag.

Reviewer 2: `wpauto` => `wpautop`
html => HTML

Author: I changed the spellings

-- **filter_normalize_style_attrs()**

`wagtail_wordpress_import/prefilters/normalize_styles_filter.py`

> This filter converts inline style attributes to a format we can work with in the next filter because the format of style attributes can be different in some for each tag. It deals with word-case, spacing and closing `;`

Reviewer 2: Wondering if there's any reason this is a separate filter?

Author: I can only think I did it that way so it could be set as not used in the config. Off the top of my head Im sure it could be part of fix_styles though and runs there.

Reviewer: **Issue**: This situation is changed in #48, making this and `filter_fix_styles` documentation out of date.

-- **filter_fix_styles()**

`wagtail_wordpress_import/prefilters/fix_styles_filter.py`

This filter applies any inline style to the html tag that is useful in the context of the Wagtail rich text editor. It performs wrapping a tag around a piece of text as well ass applying classes to a tag to represent alignment.

Examples:

> - `font-weigh:bold;` is better represented using a `<b>` tag

Reviewer 2: `font-weigh:bold;` => `font-weight: bold;`

Author: I have added the space here. But just to note the normalize_styles_filter.py does actually exclude the space when it does it's thing at the moment when it adds the style string back in.

> - `float: left;` is better represented by a class of `left-align`

Reviewer 2: I think rather than saying "is better represented by", could we word this as "is converted into"?

Author: I'll change this but realise there's inconsistency around the style spacing above, float: left should have read float:left but i'll leave it as is for now.

-- **filter_bleach_clean()**

`wagtail_wordpress_import/prefilters/bleach_filter.py`

> This filter is run last and removes tags based on a range of `ALLOWED` Tags, Attributes and Styles.

Reviewer 2: Wondering why allowed is capitalised here? Is this a setting? If so, please could you add an example?

Author: I added a link and adjusted the references so they look like settings.

---

## Pre-filter running order

> The order the pre-filters are run will have an impact on the final html output. The default order is:

Reviewer 2: html => HTML

>     1. filter_linebreaks_wp()

Reviewer 2: This is already documented here: https://github.com/torchbox/wagtail-wordpress-import/pull/36/files#diff-16d24c203df2c01d2eab0a40ebf8a1f3e693e4a6cb03e79699664b8366a30150R38

Author: I have removed this duplicate list

    2. filter_normalize_style_attrs()
    3. filter_fix_styles()
    4. filter_bleach_clean()

Each pre-filter takes the output from the previous pre-filter, with the exception of the first pre-filter which receives the raw content. 

The output of the last pre-filter is used to build the stream field blocks in a later process.

## Running order: 
It's possible to change the pre-filter running order by altering the order of the `DEFAULT_PREFILTERS` provided by adding `WAGTAIL_WORDPRESS_IMPORT_PREFILTERS` to your  own settings file.

Some pre-filters can be given `OPTIONS` to alter the way the pre-filter behaves.

## Configuration

You can use the package without providing any configuration for pre-filters. We provide a small number of useful pre-filters that process html content before it's used to build out the stream field blocks.

## Defaults

**Pre-filter Default**:

```python

> DEFAULT_PREFILTERS = [

Reviewer 2: I think ``DEFAULT_PREFILTERS`` is an implementation detail. We probably should write this something like:

By default, ``WAGTAIL_WORDPRESS_IMPORT_PREFILTERS`` is set to the following:

```
WAGTAIL_WORDPRESS_IMPORT_PREFILTERS = [
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.linebreaks_wp_filter.filter_linebreaks_wp",
    },
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.normalize_styles_filter.filter_normalize_style_attrs",
    },
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.fix_styles_filter.filter_fix_styles",
    },
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.bleach_filter.filter_bleach_clean",
    },
]
```

This way it's in a state where someone could copy it out of the docs and tweak it

Reviewer 2: I think these module paths are a bit long, would it be possible to re-export them from ``wagtail_wordpress_import.prefilters`` to make them shorter and more memorable? For example:

```
WAGTAIL_WORDPRESS_IMPORT_PREFILTERS = [
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.linebreaks_wp",
    },
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.normalize_style_attrs",
    },
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.fix_styles",
    },
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.bleach_clean",
    },
]
```

Also, think we could pick a better name for ``fix_styles``. How about ``styles_to_tags``?

Author: I'll create an issue for this: https://github.com/torchbox/wagtail-wordpress-import/issues/39

Reviewer: **Issue**: there's a discrepancy between `WAGTAIL_WORDPRESS_IMPORT_PREFILTERS` in the preceding paragraph, and the variable name of this settings dict.

    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.linebreaks_wp_filter.filter_linebreaks_wp",
    },
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.normalize_styles_filter.filter_normalize_style_attrs",
    },
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.fix_styles_filter.filter_fix_styles",
    },
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.bleach_filter.filter_bleach_clean",
    },
]
```

**To change the configuration:**

> copy the `DEFAULT_PREFILTERS` above to your own settings file and change the name to 

Reviewer 2: As above, I think we should change the example to not even mention ``DEFAULT_PREFILTERS`` since this looks like an implementation detail to me

Author: I have altered this now

    },
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.normalize_styles_filter.filter_normalize_style_attrs",
    },
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.fix_styles_filter.filter_fix_styles",
    },
    {
        "FUNCTION": "wagtail_wordpress_import.prefilters.bleach_filter.filter_bleach_clean",
        "OPTIONS": {
            "ADDITIONAL_ALLOWED_TAGS": ["my-custom-tag"]
        }
    },
]
```
`my-custom-tag` would be appended to the `ALLOWED_TAGS` in the bleach_filters

or provide your own pre-filter add:

```python
{
    "FUNCTION": "[app].[file].[method]",
},
```
at the position to match the running order you need. `[app].[file].[method]` is the  dotted path to your function.

**If you don't need any of the provided pre-filters you can exclude them from the config**

## Create pre-filters

You can find the provided pre-filters [here](wagtail_wordpress_import/prefilters)

To create your own pre-filter you need to create a function in your app with the following signature:

```python
def filter_func(input_content, options=None):
    """
    params: 
    
    input_content

>     most likely a string but depending on your implementation 

Reviewer 2: Do we allow pre-filters to be used on other types?

Author: I have changed this

    could be any type that you need to work with

    options:

>     you probably won't need to pass in any options 

Reviewer 2: Could you describe what it's for?

Author: I have added an example
